packages/flutter_tools/lib/src/platform_plugins.dart

```diff
@@ -0,0 +1,118 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+import 'package:yaml/yaml.dart';
+
+/// Marker interface for all platform specific plugin config impls.
+abstract class PlatformPlugin {
+  const PlatformPlugin();
+
+  Map<String, dynamic> toMap();
+}
+
+/// Contains parameters to template an Android plugin.
+///
+/// The required fields include: [name] of the plugin, [package] of the plugin and
+/// the [pluginClass] that will be the entry point to the plugin's native code.
+class AndroidPlugin extends PlatformPlugin {
+  const AndroidPlugin({
+    @required this.name,
+    @required this.package,
+    @required this.pluginClass,
+  });
+
+  factory AndroidPlugin.fromYaml(String name, YamlMap yaml) {
+    assert(yaml != null);
+    return AndroidPlugin(
+      name: name,
+      package: yaml['package'],
+      pluginClass: yaml['pluginClass'],
+    );
+  }
+
+  static const String kConfigKey = 'android';
+
+  final String name;
+  final String package;
+  final String pluginClass;
+
+  @override
+  Map<String, dynamic> toMap() {
+    return <String, dynamic>{
+      'name': name,
+      'package': package,
+      'class': pluginClass,
+    };
+  }
+}
+
+/// Contains the parameters to template an iOS plugin.
+///
+/// The required fields include: [name] of the plugin, the [pluginClass] that
+/// will be the entry point to the plugin's native code. [classPrefix] is required
+/// if the plugin is using Objective-C, it is not required for Swift based iOS plugins.
+class IOSPlugin extends PlatformPlugin {
+  const IOSPlugin({
+    @required this.name,
+    this.classPrefix,
+    @required this.pluginClass,
+  });
+
+  factory IOSPlugin.fromYaml(String name, YamlMap yaml) {
+    assert(yaml != null);
+    return IOSPlugin(
+      name: name,
+      classPrefix: yaml['classPrefix'],
+      pluginClass: yaml['pluginClass'],
+    );
+  }
+
+  static const String kConfigKey = 'ios';
+
+  final String name;
+  final String classPrefix;
+  final String pluginClass;
+
+  @override
+  Map<String, dynamic> toMap() {
+    return <String, dynamic>{
+      'name': name,
+      'prefix': classPrefix,
```

Comment:
We can't AFAICT; see [this comment](https://github.com/flutter/flutter/pull/38632#discussion_r315001504)

Reply:
Though we won't be breaking any existing code if we do it right now, we would make the migration that have to happen anyway harder (I'd think not by much).

On a related note - Stuart do you think it would be useful to have a (potentially optional) `fileName` parameter here and for macOS?

Reply:
> Though we won't be breaking any existing code if we do it right now, we would make the migration that have to happen anyway harder (I'd think not by much).
I don't follow. If you remove the parameter, then it's impossible to describe existing ObjC plugins using the new format. Are you suggesting we require everyone to rename their source files when migrating?

> On a related note - Stuart do you think it would be useful to have a (potentially optional) `fileName` parameter here and for macOS?

It's not clear to me how replacing `classPrefix` with `fileName` helps; it seems like the net effect is that we are still stuck with an extra parameter that we would rather not have, but with the added downside that migration between old and new formats is less obvious.

I don't want it for macOS; I can't see any compelling non-legacy reason to support custom filenames for source files when the convention of filename=class name is simple, and widely used in general.

Reply:
> > Though we won't be breaking any existing code if we do it right now, we would make the migration that have to happen anyway harder (I'd think not by much).
> 
> I don't follow. If you remove the parameter, then it's impossible to describe existing ObjC plugins using the new format. Are you suggesting we require everyone to rename their source files when migrating?

Yes.

> 
> > On a related note - Stuart do you think it would be useful to have a (potentially optional) `fileName` parameter here and for macOS?
> 
> It's not clear to me how replacing `classPrefix` with `fileName` helps; it seems like the net effect is that we are still stuck with an extra parameter that we would rather not have, but with the added downside that migration between old and new formats is less obvious.
> 
> I don't want it for macOS; I can't see any compelling non-legacy reason to support custom filenames for source files when the convention of filename=class name is simple, and widely used in general.

Makes sense, I didn't have a preference here, just wanted to check because if we would want to eventually support that we should not introduce extra migration pain right now.

Reply:
If you're okay with including a file rename step in the migration for most existing plugins, then I would agree with removing classPrefix. (The plugin template for language=objc would need to be updated accordingly.) That would need to be prominently documented somewhere.

Reply:
Let's start without `classPrefix`, when this PR makes it to stable we essentially start a "soft migration" period (as we still support the old schema). If we learn that the migration is too painful we could later add `classPrefix` (it shouldn't be breaking for plugins that already migrated without `classPrefix`).

We will definitely need documentation and announcements around this migration.

Reply:
> Let's start without `classPrefix`

Then we'll need to change the `objc` plugin template (even though it's no longer the default, it can still be selected) to create the files with the prefixes included in the same PR that changes the `pubspec.yaml` in the template. (Which doesn't need to be this PR, but if it's not going to be, then let's get an issue tracking that so it doesn't get forgotten.)